Add WIF documentation for Azure Auth and Secrets engines (#27185)
hashicorp/vault: website/content/docs/secrets/azure.mdx

   If you are running Vault inside an Azure VM with MSI enabled, `client_id` and
   `client_secret` may be omitted. For more information on authentication, see the [authentication](#authentication) section below.

   In some cases, you cannot set sensitive account credentials in your
   Vault configuration. For example, your organization may require that all
   security credentials are short-lived or explicitly tied to a machine identity.

   To provide managed identity security credentials to Vault, we recommend using Vault
   [plugin workload identity federation](#plugin-workload-identity-federation-wif)
   (WIF) as shown below.

1.  Alternatively, configure the audience claim value and the Client, Tenant and Subscription IDs for plugin workload identity federation:

   ```text
   $ vault write azure/config \
       subscription_id=$AZURE_SUBSCRIPTION_ID \
       tenant_id=$AZURE_TENANT_ID \
       client_id=$AZURE_CLIENT_ID \
       identity_token_audience=$TOKEN_AUDIENCE
   ```

>    Vault's identity token provider will internally sign the plugin identity token JWT.
>    Given a trust relationship is configured between Vault and Azure via
>    Workload Identity Federation, the secrets engine can exchange this identity token to obtain
>    federated access tokens.
> 
>    ~> **Notice:** For this trust relationship to be established, Azure must have an
>    a [Federated Identity Credential](https://learn.microsoft.com/en-us/entra/workload-id/workload-identity-federation-create-trust?pivots=identity-wif-apps-methods-azp#configure-a-federated-identity-credential-on-an-app)
>    configured with information about the fully qualified and network-reachable
>    Issuer URL for Vault's plugin [identity token provider](/vault/api-docs/secret/identity/tokens#read-plugin-identity-well-known-configurations).
>    This is to ensure that Azure can fetch the JWKS [public keys](/vault/api-docs/secret/identity/tokens#read-active-public-keys)
>    and verify the plugin identity token signature. To configure Vault's Issuer,
>    please refer to the Identity Tokens
>    [documentation](/vault/api-docs/secret/identity/tokens#configure-the-identity-tokens-backend)

Reviewer: ```suggestion
   The Vault identity token provider signs the plugin identity token JWT internally.
   If a trust relationship exists between Vault and Azure through WIF, the auth
   engine can exchange the Vault identity token for a federated access token.
     
   To configure a trusted relationship between Vault and Azure, :

      - You must configure the [identity token issuer backend](/vault/api-docs/secret/identity/tokens#configure-the-identity-tokens-backend)
        for Vault.

      - Azure must have a
        [federated identity credential](https://learn.microsoft.com/en-us/entra/workload-id/workload-identity-federation-create-trust?pivots=identity-wif-apps-methods-azp#configure-a-federated-identity-credential-on-an-app)
        configured with information about the fully qualified and network-reachable
        issuer URL for the Vault plugin
        [identity token provider](/vault/api-docs/secret/identity/tokens#read-plugin-identity-well-known-configurations).

   Establishing a trusted relationship between Vault and Azure ensures that Azure
   can fetch JWKS
   [public keys](/vault/api-docs/secret/identity/tokens#read-active-public-keys)
   and verify the plugin identity token signature.
```

|------------------------------------------------| ------------ | ------------------------------------------- |
| [User Access Administrator][user_access_admin] | Subscription | Service Principal ID given in configuration |

## Plugin Workload Identity Federation (WIF)

<EnterpriseAlert product="vault" />

The Azure secrets engine supports the Plugin WIF workflow, and has a source of identity called
a plugin identity token. The plugin identity token is a JWT that is internally signed by Vault's
[plugin identity token issuer](/vault/api-docs/secret/identity/tokens#read-plugin-workload-identity-issuer-s-openid-configuration).

If there is a trust relationship configured between Vault and Azure through

> [Workload Identity Federation](https://learn.microsoft.com/en-us/entra/workload-id/workload-identity-federation),

Reviewer: ```suggestion
[workload identity federation](https://learn.microsoft.com/en-us/entra/workload-id/workload-identity-federation),
```

   if you need to limit Vault API exposure.

1. Configure a
   [federated identity credential](https://learn.microsoft.com/en-us/entra/workload-id/workload-identity-federation-create-trust?pivots=identity-wif-apps-methods-azp#configure-a-federated-identity-credential-on-an-app)
   on a dedicated application registration in Azure to establish a trust relationship with Vault.
   1. The issuer URL **must** point at your [Vault plugin identity token issuer](/vault/api-docs/secret/identity/tokens#read-plugin-workload-identity-issuer-s-openid-configuration) with the
   `/.well-known/openid-configuration` suffix removed. For example:
   `https://host:port/v1/identity/oidc/plugins`.
   1. The subject identifier **must** match the unique `sub` claim issued by plugin identity tokens.
   The subject identifier should have the form `plugin-identity:<NAMESPACE>:secret:<AZURE_MOUNT_ACCESSOR>`.
   1. The audience should be under 600 characters. The default value in Azure is `api://AzureADTokenExchange`.

1. Configure the Azure secrets engine with the subscription, client and tenant IDs and the OIDC audience value.

> ```shell-session
> $ vault write azure/config \
>     subscription_id=$AZURE_SUBSCRIPTION_ID \
>     tenant_id=$AZURE_TENANT_ID \
>     client_id=$AZURE_CLIENT_ID \
>     identity_token_audience="vault.example/v1/identity/oidc/plugins"
> ```

Reviewer 2: ````suggestion
   ```shell-session
   $ vault write azure/config \
       subscription_id=$AZURE_SUBSCRIPTION_ID \
       tenant_id=$AZURE_TENANT_ID \
       client_id=$AZURE_CLIENT_ID \
       identity_token_audience="vault.example/v1/identity/oidc/plugins"
   ```
````
